centreon/src/Core/ResourceAccess/Application/UseCase/UpdateRule/UpdateRule.php

```diff
@@ -324,34 +327,129 @@ private function updateResourceLinks(UpdateRuleRequest $updateRequest): void
+                                datasetId: $datasetId,
+                                resourceType: $parentApplicableFilter->getType(),
+                                resourceIds: $parentApplicableFilter->getResourceIds()
+                            );
+                        }
+                    }
+                }
+
+                if ($this->shouldUpdateDatasetAccesses($applicableFilter)) {
+                    $this->updateDatasetAccesses(
+                        datasetId: $datasetId,
+                        resourceType: $applicableFilter->getType()
+                    );
+                } else {
                     $this->writeRepository->linkResourcesToDataset(
-                        $updateRequest->id,
-                        $datasetId,
-                        $parentApplicableFilter->getType(),
-                        $parentApplicableFilter->getResourceIds()
+                        ruleId: $updateRequest->id,
+                        datasetId: $datasetId,
+                        resourceType: $applicableFilter->getType(),
+                        resourceIds: $applicableFilter->getResourceIds()
                     );
                 }
-
-                // link resources to the dataset
-                $this->writeRepository->linkResourcesToDataset(
-                    $updateRequest->id,
-                    $datasetId,
-                    $applicableFilter->getType(),
-                    $applicableFilter->getResourceIds()
-                );
             }
 
             $index++;
         }
     }
 
+    /**
+     * @param DatasetFilter $parent
+     * @param DatasetFilter $child
+     */
+    private function shouldBothFiltersBeSaved(DatasetFilter $parent, DatasetFilter $child): bool
+    {
+        return DatasetFilter::isGroupOrCategoryFilter($child)
+            && DatasetFilter::isGroupOrCategoryFilter($parent);
+    }
+
+    /**
+     * @param int $datasetId
+     * @param string $resourceType
+     */
+    private function updateDatasetAccesses(int $datasetId, string $resourceType): void
+    {
+        switch ($resourceType) {
+            case HostGroupFilterType::TYPE_NAME:
+                $this->writeRepository->updateDatasetAccess(
+                    datasetId: $datasetId,
+                    resourceType: 'hostgroups',
+                    fullAccess: true
+                );
+                break;
+            case ServiceGroupFilterType::TYPE_NAME:
+                $this->writeRepository->updateDatasetAccess(
+                    datasetId: $datasetId,
+                    resourceType: 'servicegroups',
+                    fullAccess: true
+                );
+                break;
+            case HostFilterType::TYPE_NAME:
+                $this->writeRepository->updateDatasetAccess(
+                    datasetId: $datasetId,
+                    resourceType: 'hosts',
+                    fullAccess: true
+                );
+                break;
+        }
```

Comment:
Phpstan not whining about no default case ?

Reply:
Nope. 

<img width="1117" alt="image" src="https://github.com/centreon/centreon/assets/31647811/3e954469-1b70-4834-9901-7bea549829c4">